src/Features/CSharp/Portable/ForeachToFor/CSharpForeachToForCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,160 @@
+            if (!scope.IntersectsWith(token.Span))
+            {
+                return null;
+            }
+
+            // check whether there is any comments between foreach and ) tokens
+            // if they do, we don't support conversion.
+            foreach (var trivia in foreachStatement.DescendantTrivia(n => n == foreachStatement || scope.Contains(n.FullSpan)))
+            {
+                if (trivia.Span.End <= scope.Start ||
+                    scope.End <= trivia.Span.Start)
+                {
+                    continue;
+                }
+
+                if (trivia.Kind() != SyntaxKind.WhitespaceTrivia &&
+                    trivia.Kind() != SyntaxKind.EndOfLineTrivia)
+                {
+                    // we don't know what to do with these comments
+                    return null;
+                }
+            }
+
+            return foreachStatement;
+        }
+
+        protected override (SyntaxNode start, SyntaxNode end) GetForeachBody(SyntaxNode node)
+        {
+            var foreachStatement = (ForEachStatementSyntax)node;
+            return (foreachStatement.Statement, foreachStatement.Statement);
+        }
+
+        protected override void ConvertToForStatement(SemanticModel model, ForeachInfo foreachInfo, SyntaxEditor editor, CancellationToken cancellationToken)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+
+            var parseOption = model.SyntaxTree.Options;
+            var foreachStatement = (ForEachStatementSyntax)foreachInfo.ForeachStatement;
+
+            // first, see whether we need to introduce new statement to capture collection
+            var foreachCollectionExpression = GetForeachCollection(foreachStatement);
+            var collectionVariableName = foreachCollectionExpression.ToString();
+            if (foreachInfo.RequireCollectionStatement)
+            {
+                collectionVariableName = CreateUniqueMethodName(model, foreachStatement, "list");
+
+                var collectionStatement = SyntaxFactory.LocalDeclarationStatement(
+                    SyntaxFactory.VariableDeclaration(
+                        SyntaxFactory.IdentifierName("var"),
+                        SyntaxFactory.SingletonSeparatedList(
+                            SyntaxFactory.VariableDeclarator(
+                                SyntaxFactory.Identifier(collectionVariableName).WithAdditionalAnnotations(RenameAnnotation.Create()),
+                                default,
+                                SyntaxFactory.EqualsValueClause(
+                                    foreachInfo.RequireExplicitCast ?
+                                        SyntaxFactory.CastExpression(
+                                            SyntaxFactory.ParseTypeName(foreachInfo.ExplicitCastInterface).WithAdditionalAnnotations(Simplifier.Annotation),
```

Comment:
Please dont' parse type names.  We can always go from a type symbol to a TypeSyntax using existing helpers.

Reply:
I am not sure why you dont want to parse? there is nothing wrong parsing snippet of code rather than building them using SyntaxFactory.

Reply:
Because it does not do the right thigns (like add the appropriate annotations).  Parsing means your'e dependent on what ToDisplayString produces.  it means that things like "AddUsing" don't work properly.  And whatnot.  

We already built entire APIs to support this scenario and we use them in *all* our refactorings and analyzers.  Please do not create code that deviates from the patterns and practices of hte IDE for no reason. 

These APIs work well and they do the right thing wrt to the entire code-action stack.  They put the right simplification annotations in place and they are where we can go and fix things such that they then apply to the entire codebase.

For example, if you use .LocalDeclarationStatement then that is a place where we can update it to use 'var' or an explicit-type, and then all features benefit.

Reply:
> there is nothing wrong

Yes there is.  features that are coded differently and work differently than the rest of the IDE act as a tax on future maintainers.  It means that instead of having uniform patterns and approaches to solving problems, the maintainer has to figure out why things are done differently.  That's why we create supplementary APIs to address these common patterns, and why we want all our features to use them.  please do not be different here. This feature is not special and does not warrant side-stepping the infrastructure we've spent over a decade building.  

This is the philosophy we take with all our code action features (regardless of whether they are internally or externally contributed).

Reply:
and that's why I telling people it is completely fine parsing some snippet rather than creating this verbose not intuitive chunk of tree building.

Reply:
> And that's why I telling people it is completely fine parsing some snippet 

I literally listed several reasons above why it is not "completely fine".  You yourself encountered one of those reasons yourself when your parsed snippet wouldn't format properly because it didn't contain the right trivia on it that the rest of our stack expects and works with.

Again, 100% of roslyn IDE code works this way.  This is the way we recommend to users that things be done.  It's the way that ensures that our stack works as uniformly and consistently as possible.  Please, do not deviate from the rest of the IDE code.  It is deliberate and intentional how it works, and deviations just end up being bugs and costs down hte line for whoever has to maintain 'ForEachToFor'.